torchao/prototype/autoround/core.py

```diff
@@ -0,0 +1,269 @@
+import dataclasses
+import logging
+from typing import Callable, Dict, Optional, Tuple
+
+import torch
+from torch.utils._pytree import tree_flatten, tree_unflatten
+
+import torchao.prototype.autoround.utils as ar_utils
+import torchao.quantization as ao_quant
+from torchao.dtypes import TensorCoreTiledLayoutType, to_affine_quantized_static
+from torchao.prototype.autoround.multi_tensor import _multi_tensor_config, MultiTensor
+from torchao.quantization.quant_primitives import ZeroPointDomain
+from torchao.utils import find_multiple
+
+
+@ar_utils.singleton
+@dataclasses.dataclass
+class _AutoRoundConfig:
+    bits: int = 4
+    group_size: int = 128
+    iters: int = 200
+
+
+_auto_round_config = _AutoRoundConfig()
+
+
+@ar_utils.singleton
+@dataclasses.dataclass
+class _OptimizationTracker:
+    num_layers: int = 0
+    optimized_layers: int = 0
+
+
+_optimization_tracker = _OptimizationTracker()
+
+
+@torch.no_grad()
+def prepare_model_for_applying_auto_round_(
+    model: torch.nn.Module,
+    is_target_module: Callable[[torch.nn.Module, str], bool],
+    bits: int = 4,
+    group_size: int = 128,
+    iters: int = 200,
+    device: Optional[torch.types.Device] = None,
+):
+
+    _multi_tensor_config.accelerator_name = device
+
+    _auto_round_config.bits = bits
+    _auto_round_config.group_size = group_size
+    _auto_round_config.iters = iters
+
+    def forward_hook(
```

Comment:
actually I'm wondering if this should be `forward_pre_hook` instead, I also updated my example with that: https://github.com/pytorch/ao/pull/721

Reply:
Thanks for clarifying. However, I prefer to continue using `forward_hook` for `auto-round`, as it requires the output of the module to calculate reconstruction loss.